spacy/ml/models/span_predictor.py

```diff
@@ -0,0 +1,238 @@
+from typing import List, Tuple
+
+from thinc.api import Model, chain, tuplify
+from thinc.api import PyTorchWrapper, ArgsKwargs
+from thinc.types import Floats2d, Ints1d
+from thinc.util import torch, xp2torch, torch2xp
+
+from ...tokens import Doc
+from ...util import registry
+from .coref_util import get_sentence_ids
+
+
+@registry.architectures("spacy.SpanPredictor.v1")
+def build_span_predictor(
+    tok2vec: Model[List[Doc], List[Floats2d]],
+    hidden_size: int = 1024,
+    distance_embedding_size: int = 64,
+    conv_channels: int = 4,
+    window_size: int = 1,
+    max_distance: int = 128,
+    prefix: str = "coref_head_clusters",
+):
+    # TODO add model return types
+    # TODO fix this
+    try:
+        dim = tok2vec.get_dim("nO")
+    except ValueError:
+        # happens with transformer listener
+        dim = 768
+
+    with Model.define_operators({">>": chain, "&": tuplify}):
+        span_predictor = PyTorchWrapper(
+            SpanPredictor(
+                dim,
+                hidden_size,
+                distance_embedding_size,
+                conv_channels,
+                window_size,
+                max_distance,
+            ),
+            convert_inputs=convert_span_predictor_inputs,
+        )
+        # TODO use proper parameter for prefix
+        head_info = build_get_head_metadata(prefix)
+        model = (tok2vec & head_info) >> span_predictor
+
+    return model
+
+
+def convert_span_predictor_inputs(
+    model: Model, X: Tuple[Ints1d, Tuple[Floats2d, Ints1d]], is_train: bool
+):
+    tok2vec, (sent_ids, head_ids) = X
+    # Normally we should use the input is_train, but for these two it's not relevant
+
+    def backprop(args: ArgsKwargs) -> List[Floats2d]:
+        gradients = torch2xp(args.args[1])
+        return [[gradients], None]
+
+    word_features = xp2torch(tok2vec[0], requires_grad=is_train)
+    sent_ids = xp2torch(sent_ids[0], requires_grad=False)
+    if not head_ids[0].size:
+        head_ids = torch.empty(size=(0,))
+    else:
+        head_ids = xp2torch(head_ids[0], requires_grad=False)
+
+    argskwargs = ArgsKwargs(args=(sent_ids, word_features, head_ids), kwargs={})
+    return argskwargs, backprop
```

Comment:
I don't understand the code well enough yet to understand the typing here.

The `backprop` call promises `List[Floats2d]` but produces `[[gradients], None]`. What is the `None` for?

Reply:
Also, `tok2vec[0]` is just one single `int`, no? That's run through `xp2torch` and stored in `word_features`?

Reply:
So the types are definitely weird here for a few reasons.

This code was originally written to handle batching. It was re-written to only handle fake batching, so that while the input and output look like they contain multiple items, they can only actually have info for one doc. 

I believe that since every implementation we've seen of similar architectures doesn't actually do batching we decided this was OK. This can definitely be cleaned up, but because it's kind of weird I think the final code is still going to be odd looking.